c7n/resources/batch.py

```diff
@@ -41,3 +43,109 @@ class resource_type(object):
         id = name = "jobDefinitionName"
         enum_spec = (
             'describe_job_definitions', 'jobDefinitions', None)
+
+
+class StateTransitionFilter(object):
+    """Filter resources by state.
+
+    Try to simplify construction for policy authors by automatically
+    filtering elements (filters or actions) to the resource states
+    they are valid for.
+    """
+    valid_origin_states = ()
+
+    def filter_resource_state(self, resources, key, states=None):
+        states = states or self.valid_origin_states
+        if not states:
+            return resources
+        orig_length = len(resources)
+        results = [r for r in resources if r[key] in states]
+        self.log.info("%s %d of %d resources with valid %s" % (
+            self.__class__.__name__, len(results), orig_length, key.lower()))
+        return results
+
+
+@ComputeEnvironment.action_registry.register('update-environment')
+class UpdateComputeEnvironment(BaseAction, StateTransitionFilter):
+    """Updates an AWS batch compute environment
+
+    :example:
+
+    .. code-block: yaml
+
+        policies:
+          - name: update-environments
+            resource: batch-compute
+            filters:
+              - computeResources.desiredvCpus: 0
+              - state: ENABLED
+            actions:
+              - type: update-environment
+                state: DISABLED
+    """
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['update-environment']},
+            'computeEnvironment': {'type': 'string'},
+            'state': {'type': 'string', 'enum': ['ENABLED', 'DISABLED']},
+            'computeResources': {
+                'type': 'object',
+                'additionalProperties': False,
+                'properties': {
+                    'minvCpus': {'type': 'integer'},
+                    'maxvCpus': {'type': 'integer'},
+                    'desiredvCpus': {'type': 'integer'}
+                }
+            },
+            'serviceRole': {'type': 'string'}
+        }
+    }
+    permissions = ('batch:UpdateComputeEnvironment',)
+    valid_origin_status = ('VALID', 'INVALID')
+
+    def process(self, resources):
+        resources = self.filter_resource_state(
+            resources, 'status', self.valid_origin_status)
```

Comment:
if we filter all the resources (ie if not resources: ) it would be good to log to that effect that we're implicitly doing filter in the action.